src/config-file-watcher.ts

```diff
@@ -0,0 +1,78 @@
+import * as path from 'path';
+import * as vscode from 'vscode';
+
+import * as config from './config';
+
+export function activate(context: vscode.ExtensionContext, args: string[]) {
+  let compileCommandsFolder: string|undefined = undefined;
+
+  for (let arg of args) {
+    if (arg.startsWith('--compile-commands-dir')) {
+      let match = arg.match(/--compile-commands-dir="?([^"]*)"?/);
+      compileCommandsFolder = match[1] || undefined;
+    }
+  }
+
+  const watcher = new ConfigFileWatcher(context, compileCommandsFolder);
+}
+
+class ConfigFileWatcher {
+  private databaseWatcher = vscode.workspace.createFileSystemWatcher(
+      '**/{compile_commands.json,compile_flags.txt,.clang-tidy}');
+
+  constructor(private context: vscode.ExtensionContext,
+              private compileCommandsFolder: string|undefined) {
+    this.databaseWatcher.onDidChange(this.handleConfigFilesChanged.bind(this));
+    this.databaseWatcher.onDidCreate(this.handleConfigFilesChanged.bind(this));
+    context.subscriptions.push(this.databaseWatcher);
+  }
+
+  async handleConfigFilesChanged(uri: vscode.Uri) {
+    if ((await vscode.workspace.fs.readFile(uri)).length) {
```

Comment:
It shouldn't read the whole file for that either.
```suggestion
    if ((await vscode.workspace.fs.stat(uri)).size) {
```

Reply:
Also, early exit!
```suggestion
    if ((await vscode.workspace.fs.stat(uri)).size <= 0)
      return;
```

Reply:
@hokein this is to check, as you understood, if the file is empty, because I have noticed that sometimes the tools that generate the compile_commands.json file, before writing to it they clear it or they delete it and create a new empty one, and this should not toggle a restart of the server.

@Trass3r much better solution thanks!